projects/transformers/callbacks/sparsity.py

```diff
@@ -75,3 +93,266 @@ def on_step_end(self, args, state, control, model, **kwargs):
+        self.sparse_modules = None
+
+    def on_init_end(self, args, state, control, model=None, **kwargs):
+        """
+        Save a list of all the sparse modules and a dict of their initial densities.
+        """
+        self.sparse_modules = filter_modules(model, include_modules=[SparseWeightsBase])
+        for name, module in self.sparse_modules.items():
+            self.initial_on_params[name] = (module.zero_mask.bool() == 0).sum().item()
+        initial_sparsity = getattr(args, "config_kwargs", {}).get("sparsity", 0)
+        self.initial_density = 1 - initial_sparsity
+
+    def on_step_end(self, args, state, control, model=None, **kwargs):
+        """
+        Plot the densities of each layer and plot the change in on-params of each layer.
+        """
+
+        if state.global_step % self.plot_freq != 0:
+            return
+
+        # Plot densities for each layer.
+        df_dendity_by_layer = get_density_by_layer(self.sparse_modules)
+        fig, ax = plt.subplots(figsize=(8, 5), constrained_layout=True)
+        sns.stripplot(
+            data=df_dendity_by_layer,
+            y="density",
+            x="layer",
+            hue=None,
+            color="firebrick",
+            ax=ax,
+        )
+        ax.axhline(self.initial_density, label="Initial Density")
+        ax.legend(loc="lower center", bbox_to_anchor=(0.2, 0), ncol=2)
+        ax.set_title("Density Per Layer")
+        ax.set_ylim(0, 1)
+        ax.set_xticklabels(
+            ax.get_xticklabels(),
+            rotation=-45,
+            ha="left",
+            rotation_mode="anchor"
+        )
+        plot = wandb.Image(ax)
+        wandb.log({"density_per_layer": plot}, step=state.global_step)
+
+        # Plot plot change in on params for each layer.
+        df_delta_on_params = get_delta_on_params(
+            self.sparse_modules,
+            self.initial_on_params
+        )
+        fig, ax = plt.subplots(figsize=(8, 5), constrained_layout=True)
+        sns.stripplot(
+            data=df_delta_on_params,
+            y="delta_on_params",
+            x="layer",
+            hue=None,
+            color="firebrick",
+            ax=ax,
+        )
+        ax.axhline(0)
+        ax.set_title("Change in On-Params Per Layer")
+        ax.set_ylabel("delta on-params")
+        ax.set_xticklabels(
+            ax.get_xticklabels(),
+            rotation=-45,
+            ha="left",
+            rotation_mode="anchor",
+        )
+        plot = wandb.Image(ax)
+        wandb.log({"delta_on_params": plot}, step=state.global_step)
+
+
+# -------------
+# Utilities
+# -------------
+
+def inputs_to_device(inputs, device):
+    """
+    Prep inputs dict by transferring all inputs to the device.
+    """
+    for k, v in inputs.items():
+        if isinstance(v, torch.Tensor):
+            inputs[k] = v.to(device)
+
+
+def calc_sparsity(tensor):
+    """Calculate the sparsity of a given tensor."""
+    num_zero = (tensor == 0).sum().item()
```

Comment:
For all cases where we have `t == 0` or `t != 0`, a maybe faster alternative is `torch.count_nonzero`

Reply:
I've made the change in case it helps.